### PR TITLE
Show warning when server time is off from device time

### PR DESCRIPTION
**Pull request by dechov**
#### Changes proposed in this Pull Request:

There have been a few cases (that I've seen, possibly far more) where an incorrect server time leads to webhook verification failure due to the timestamp check [[code](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/19e5f4a47d4ad94d4d367a588b3cd4b227c9f80d/includes/class-wc-stripe-webhook-handler.php#L104-L108)].

We are not currently communicating the need for a correct server time, and there is an opportunity to present messaging around it in the context of the "webhook secret" setting (the presence of which triggers verification).

This change is to **show a warning** when the secret is set (or is being input) in the event that the **server time and device time are more than a few minutes apart**.

Screen Shot 2020-06-02 at 2 13 20 PM

The threshold here (4 minutes) takes into account a) the 5 minute threshold for verification, and b) Settings page loading time. Also, device time is in no way guaranteed to be more correct than the server time, but messaging conditional on the discrepancy is likely to help in a significant portion of cases.

_To test:_ first verify that server time and device time are correct (within a minute or two), and that no warning appears below the "Webhook Secret" setting. Then set the server time to something incorrect (e.g. `date -s "Mon May 25 18:28:03 UTC 2020"`), and verify that a warning message appears whenever a value is present for the webhook secret. Should work (independently but identically) for live and test secrets.

-------------------
- [ ] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [ ] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.

### CODE REVIEW

File: assets/js/stripe-admin.js [hunk 1]
```diff
@@ -86,6 +88,25 @@ jQuery( function( $ ) {
 					} );
 				} );
 			} );
+
+			// Webhook verification checks for timestamp within 5 minutes so warn if
+			// server time is off from browser time by > 4 minutes.
+			var timeDifference = Date.now() / 1000 - wc_stripe_settings_params.time;
+			var isTimeOutOfSync = Math.abs( timeDifference ) > 4 * 60;
+			$( '#woocommerce_stripe_test_webhook_secret, #woocommerce_stripe_webhook_secret' )
+				.on( 'change input', function() {
```
Comment (allendav) [inline assets/js/stripe-admin.js:133]: Why wait to change input? If a time delta emerges later, it would still be valueable to see this in the settings page, even if they haven't changed the webhook signing secret.  How about if signing secret not empty instead?

Comment (dechov) [reply assets/js/stripe-admin.js:133]: It doesn't actually wait – see `.change()` call below. Also, the `$( this ).val()` condition below makes sure the secret is not empty.

Comment (allendav) [reply assets/js/stripe-admin.js:133]: Ah - missed that!

File: includes/class-wc-gateway-stripe.php [hunk 2]
```diff
@@ -350,7 +350,18 @@ public function admin_scripts() {
 
 		$suffix = defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG ? '' : '.min';
 
-		wp_enqueue_script( 'woocommerce_stripe_admin', plugins_url( 'assets/js/stripe-admin' . $suffix . '.js', WC_STRIPE_MAIN_FILE ), array(), WC_STRIPE_VERSION, true );
+		wp_register_script( 'woocommerce_stripe_admin', plugins_url( 'assets/js/stripe-admin' . $suffix . '.js', WC_STRIPE_MAIN_FILE ), array(), WC_STRIPE_VERSION, true );
+
+		$params = array(
+			'time'             => time(),
+			'i18n_out_of_sync' => wp_kses(
+				__( '<strong>Warning:</strong> server time is out of sync with the time on your device. Webhook verification depends on the correct time, so please <strong>check your server time</strong> before setting a webhook secret.', 'woocommerce-gateway-stripe' ),
```
Comment (allendav) [inline includes/class-wc-gateway-stripe.php]: I'm not sure your average merchant will appreciate the importance of webhook verification. Maybe reword something like this? (and this will also make it make more sense if the time gets out of sync after they've set the secret)

Warning: your site's time appears to be incorrect. Payment webhooks depend on the site time being correct and an incorrect time can result in lost sales. Please contact your hosting provider's support to correct your site's time.

Comment (dechov) [reply includes/class-wc-gateway-stripe.php]: I'm in favor of some aspects of that change, but I think we _may_ want to avoid assuming:
- correct device time
- that sales could be affected (only true for certain payment methods)
- that they would rather get the time corrected than not have a webhook secret (but then again, maybe we should always encourage signature verification)
- they cannot change the server time themselves (this one's not a big deal)

Taking another swing at it, but feel free to revise further:

> **Warning:** server time is out of sync with the time on your device, and may be incorrect. Verifying webhooks _with signing secret_ depends on the correct time on the host – and webhooks are required for payment processing in some cases – so please <strong>check your server time</strong> before setting a webhook secret (and contact your hosting provider to correct the site's time if necessary).

Comment (allendav) [reply includes/class-wc-gateway-stripe.php]: Good idea. A little more wordsmithing:

> Warning: your site's time does not match the time on your browser, and may be incorrect. Some payment methods depend on webhook verification and verifying webhooks with a signing secret depends on your site's time being correct, so please check your site's time before setting a webhook secret. You may need to contact your site's hosting provider to correct the site's time.

**Comment by allendav** [comment]
Updated to look like this now (to reproduce, either tweak your site's time as described above or hack class-wc-gateway-stripe.php near line 356 to add or substract > 300 seconds from `time()`

Screen Shot 2020-07-22 at 2 22 51 PM

**Comment by allendav** [comment]
@luizreis or @reykjalin - would one of you test this when you get a chance? Thx!

File: assets/js/stripe-admin.js [hunk 3]
```diff
@@ -86,6 +88,25 @@ jQuery( function( $ ) {
...
+					var $td = $( this ).closest( 'td' );
+					var $warning = $td.find( '.webhook_secret_time_sync_warning' );
+
+					if ( $( this ).val() ){
```
Comment (v18) [inline assets/js/stripe-admin.js]: Would be good to call out the meaning of this using a variable like
`isWebhookSecretNotEmpty =  $( this ).val().length > 0`

or something similar - that name is a bit clunky.

File: assets/js/stripe-admin.js [hunk 4]
```diff
@@ -86,6 +88,25 @@ jQuery( function( $ ) {
...
+						if ( isTimeOutOfSync && ! $warning.length ) {
```
Comment (v18) [inline assets/js/stripe-admin.js]: Could also call out the meaning of `$warning.length` using a variable like `hasWarningAlready` etc

File: includes/class-wc-gateway-stripe.php [hunk 5]
```diff
@@ -350,7 +350,18 @@ public function admin_scripts() {
 
 		$suffix = defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG ? '' : '.min';
 
-		wp_enqueue_script( 'woocommerce_stripe_admin', plugins_url( 'assets/js/stripe-admin' . $suffix . '.js', WC_STRIPE_MAIN_FILE ), array(), WC_STRIPE_VERSION, true );
+		wp_register_script( 'woocommerce_stripe_admin', plugins_url( 'assets/js/stripe-admin' . $suffix . '.js', WC_STRIPE_MAIN_FILE ), array(), WC_STRIPE_VERSION, true );
+
+		$params = array(
+			'time'             => time(),
+			'i18n_out_of_sync' => wp_kses(
```
Comment (v18) [inline includes/class-wc-gateway-stripe.php:358]: It's a bit hard to understand what `i18n_out_of_sync` means, maybe something like `message_i18n_out_of_sync` would work better?

**Comment by v18** [comment]
This is a great initiative, thank you @dechov for being proactive about the issues stemming from #1196 💯 .

Here are the goals of this PR as I understood them:
- when a merchant enters their webhook secret and their server time is off, we should let them know it won't work and suggest a fix (done)
- when a merchant is debugging why their webhooks aren't working (and they're not working due to an incorrect server time), they'll visit the stripe settings page, and see the helpful message (done)

I think the only related problem that's missing from a merchant-perspective is when they don't know their webhooks aren't working due to a misconfigured server time. In that case, they probably won't go looking on the Stripe settings page until they see a problem. I don't know if this is a problem or not, so I'm happy to add this fix and iterate when we find out more information.

I'm also not 100% sure merchants will notice it - we can rely on support request reports to find out about that.

As for the implementation itself - it's working great for me. I have a small concern, but feel free to ignore and merge without updating.

I noticed that we're relying on the implementation details of the settings page (i.e. looking for `closest td`) to add the warning message. But I couldn't think of a neater way to add the warning message. So happy to merge and fix as needed when/if the settings implementation changes.

File: includes/class-wc-gateway-stripe.php (hunk 5 above)
Comment (dechov) [reply includes/class-wc-gateway-stripe.php:358]: That's a convention from WC core, it looks like – many examples in [this module](https://github.com/woocommerce/woocommerce/blob/master/includes/class-wc-frontend-scripts.php) (and used by this extension [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/87365188784e413bf2fd66bc504637122b823a4e/assets/js/stripe-payment-request.js#L592)). Do you think it's worth changing in this case regardless?

File: assets/js/stripe-admin.js (hunk 4 above)
Comment (dechov) [reply assets/js/stripe-admin.js]: Done in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1226/commits/46a4f40f43709f934f86c2a7997e259a98a6b2ba

File: assets/js/stripe-admin.js (hunk 3 above)
Comment (dechov) [reply assets/js/stripe-admin.js]: Done in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1226/commits/46a4f40f43709f934f86c2a7997e259a98a6b2ba
